packages/core/src/ckb/epoch.ts

```diff
@@ -0,0 +1,510 @@
+import { type Bytes, type BytesLike } from "../bytes/index.js";
+import type { ClientBlockHeader } from "../client/clientTypes.js";
+import { Zero } from "../fixedPoint/index.js";
+import { type Hex, type HexLike } from "../hex/index.js";
+import { mol } from "../molecule/index.js";
+import {
+  numBeToBytes,
+  numFrom,
+  NumLike,
+  numToHex,
+  type Num,
+} from "../num/index.js";
+import { gcd } from "../utils/index.js";
+
+/**
+ * EpochLike
+ *
+ * Union type that represents any allowed input shapes that can be converted
+ * into an Epoch instance.
+ *
+ * - Object form: { integer, numerator, denominator } where each member is NumLike
+ * - Tuple form: [integer, numerator, denominator] using NumLike values
+ *
+ * All numeric-like values are converted via numFrom() when creating an Epoch.
+ */
+export type EpochLike =
+  | {
+      integer: NumLike;
+      numerator: NumLike;
+      denominator: NumLike;
+    }
+  | [NumLike, NumLike, NumLike];
+
+/**
+ * Epoch
+ *
+ * Represents a blockchain epoch composed of a whole integer part and an optional
+ * fractional part represented with numerator/denominator. This class stores
+ * values using the project's Num (bigint) type and provides utilities for
+ * packing/unpacking, arithmetic, normalization and conversion to an estimated
+ * Unix timestamp based on a reference block header.
+ *
+ * Important invariants:
+ * - For the canonical (non-zero) epoch the denominator must be > 0.
+ * - The zero epoch is represented with integer=0, numerator=0 and denominator=0.
+ * - Use normalized() to obtain a canonical representation (reduced fraction and
+ *   numerator < denominator).
+ *
+ * @example
+ * const e = new Epoch(numFrom(1), numFrom(2), numFrom(3)); // 1 + 2/3
+ */
+export class Epoch extends mol.Entity.Base<EpochLike, Epoch>() {
+  /**
+   * Construct a new Epoch instance.
+   *
+   * @param integer - Whole epoch units (Num).
+   * @param numerator - Fractional numerator (Num).
+   * @param denominator - Fractional denominator (Num). For a non-zero epoch must be > 0.
+   *                      For the zero epoch, denominator can be 0 (with integer and numerator also 0).
+   * @throws Error If denominator is zero while integer or numerator are non-zero.
+   * @throws Error If denominator is negative.
+   */
+  public constructor(
+    public readonly integer: Num,
+    public readonly numerator: Num,
+    public readonly denominator: Num,
+  ) {
+    // Ensure the epoch has a non zero in non-zero epoch.
+    if (denominator === Zero && (integer !== Zero || numerator !== Zero)) {
```

Comment:
Considering that we already accept `0 - 0/0`, `1 - 0/0` might also be acceptable. Of course, `1 - 1/0` is still too much.